lib/spack/spack/solver/asp.py

```diff
@@ -2107,8 +2108,17 @@ def build_specs(self, function_tuples):
         for s in self._specs.values():
             _develop_specs_from_env(s, ev.active_environment())
 
-        for s in self._specs.values():
-            s._mark_concrete()
+        # mark concrete and assign hashes to all specs in the solve
+        for root in roots.values():
+            # See docs for in _assign_hash for why package_hash needs to happen here.
+            root._assign_hash(ht.package_hash)
+
+            # Mark everything in the spec as concrete
+            root._mark_concrete()
+
+            # Assign dag_hash (this *could* be done lazily, but it's assigned anyway in
+            # ensure_no_deprecated, and it's clearer to see explicitly where it happens)
+            root._assign_hash(ht.dag_hash)
```

Comment:
This pattern might warrant encapsulation in a method, and then we can ban `_assign_hash` from being called anywhere else. `finalize_concretization` could be a public method, and then we wouldn't be calling private methods from another module too.